bindings/pyroot_experimental/PyROOT/src/DeclareCallablePyz.cxx

```diff
@@ -0,0 +1,663 @@
+
+      auto pytype = typemap.find(inputTypesStr[idx]);
+      if (pytype != typemap.end()) { // Types in typemap
+         pytypes[idx] = pytype->second;
+         vars << pytypes[idx] << "_" << idx;
+         code << inputTypesStr[idx] << " " << pytypes[idx] << "_" << idx;
+      } else if (inputTypesStr[idx].compare("") == 0) { // No input type
+         pytypes[idx] = "";
+      } else if (inputTypesStr[idx].compare("bool") == 0) { // Bool
+         pytypes[idx] = "O";
+         vars << "pyb_" << idx;
+         code << inputTypesStr[idx] << " b_" << idx;
+      } else { // C++ object
+         pytypes[idx] = "O";
+         vars << "pyo_" << idx;
+         code << inputTypesStr[idx] << "& o_" << idx;
+      }
+      typestr << pytypes[idx];
+
+      if (idx != inputTypesSize - 1 && pytypes[idx].compare("") != 0) {
+         code << ", ";
+         vars << ", ";
+      }
+
+      idx++;
+   }
+   Py_DECREF(iter);
+
+   // Acquire lock to protect multi-threaded scenarios
+   code << ") {\n"
+        << "   // Acquire lock to protect multi-threaded scenarios\n"
+        << "   R__WRITE_LOCKGUARD(ROOT::gCoreMutex);\n\n";
+
+   // Get pointer to Python callable
+   code << "   // Get Python callable from pointer\n"
+        << "   auto pyfunc = reinterpret_cast<PyObject*>(" << pyfunc << ");\n"
+        << "   if (!PyCallable_Check(pyfunc)) {\n"
+        << "      throw std::runtime_error(\"Python object " << name << " is not callable.\");\n"
+        << "   }\n\n";
+
+   // Build Python proxies of the C++ objects for Python
+   code << "   // Build Python proxies for C++ objects\n";
+   std::stringstream cleanup; // Register clean-up code
+   bool hasPyBoolIncref = false;
+   for (std::size_t i = 0; i < pytypes.size(); i++) {
+      if (inputTypesStr[i].compare("bool") == 0) { // Bool
+         if (!hasPyBoolIncref) {
+            code << "   Py_INCREF(Py_True);\n"
+                 << "   Py_INCREF(Py_False);\n";
+            cleanup << "   Py_DECREF(Py_True);\n"
+                    << "   Py_DECREF(Py_False);\n";
+            hasPyBoolIncref = true;
+         }
+         code << "   auto pyb_" << i << " = b_" << i << " ? Py_True : Py_False;\n";
+      } else if (pytypes[i] == "O") { // C++ objects
+         code << "   auto pyo_" << i << " = TPython::CPPInstance_FromVoidPtr("
+              << "&o_" << i << ", \"" << inputTypesStr[i] << "\");\n";
+         cleanup << "   Py_DECREF(pyo_" << i << ");\n";
+      }
+   }
+   code << "\n";
+
+   // Call Python callable
+   auto typestr_str = typestr.str();
+   auto vars_str = vars.str();
+   if (vars_str.compare("") != 0) {
+      vars_str = ", " + vars_str;
+   }
+   code << "   // Call Python callable\n"
+        << "   auto pyresult = PyObject_CallFunction(pyfunc, (char*)\"" << typestr_str << "\"" << vars_str << ");\n"
+        << "   if (pyresult == 0) {\n"
+        << "      PyErr_Print();\n"
+        << "      throw std::runtime_error(\"Failed to call Python callable " << name << ".\");\n"
```

Comment:
Is this caught by PyROOT and presented as a Python error?

Reply:
No, the Python error is printed to stdout and the exception is thrown with the set runtime error.